### PR TITLE
Fix a potential runtime in uplink implants

### PR DESCRIPTION
**Pull request by Absolucy**
## About The Pull Request

Literally just https://github.com/tgstation/tgstation/pull/63283, except I actually filled out the PR outline 2 years later.

## Why It's Good For The Game

if anyone ever implements something like my good ol' infiltration gamemode (where an outfit gave ppl uplink implants), they won't run into this error.

also less errors > more errors.

## Changelog

:cl:
fix: Fixed a potential runtime error with uplink implants. The chances of said error happen are basically zero during normal gameplay, but 0% theoretical errors are better than 0.0001% theoretical errors.
/:cl:

### CODE REVIEW

File: code/datums/components/uplink.dm [hunk 1]
```diff
@@ -330,8 +330,9 @@
 /datum/component/uplink/proc/implanting(datum/source, list/arguments)
 	SIGNAL_HANDLER
 
+	var/mob/target = arguments[1]
 	var/mob/user = arguments[2]
-	owner = user?.key
+	owner = user ? user.key : target?.key
```
Comment (ZephyrTFA) [inline code/datums/components/uplink.dm:335]:
```suggestion
	owner = user?.key || target?.key || CRASH("Uplink created without an owner!")
```

File: code/datums/components/uplink.dm [hunk 2]
```diff
@@ -330,8 +330,9 @@
 /datum/component/uplink/proc/implanting(datum/source, list/arguments)
 	SIGNAL_HANDLER
 
+	var/mob/target = arguments[1]
 	var/mob/user = arguments[2]
-	owner = user?.key
+	owner = user?.key || target?.key || CRASH("Uplink created without an owner!")
```
Comment (ZephyrTFA) [inline code/datums/components/uplink.dm:335]:
```suggestion
	owner = user?.key || target?.key
	if(!owner)
		CRASH("Uplink created without an owner!")
```
Turns out BYOND doesn't like C# exception throwing syntax

**Comment by Mothblocks** [comment]
lol you didnt even need to fill out the pr template for that pr

File: code/datums/components/uplink.dm [hunk 3]
```diff
@@ -330,8 +330,11 @@
 /datum/component/uplink/proc/implanting(datum/source, list/arguments)
 	SIGNAL_HANDLER
 
+	var/mob/target = arguments[1]
 	var/mob/user = arguments[2]
-	owner = user?.key
+	owner = user?.key || target?.key
+	if(!owner)
+		CRASH("Uplink created without an owner!")
 	if(owner && !purchase_log)
```
Comment (Mothblocks) [inline code/datums/components/uplink.dm:338]:
```suggestion
	if(!purchase_log)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
what i'd like to see more than a non-player facing changelog that im gonna remove and a fluff why is this good for the game is what the runtime actually is

**Comment by Absolucy** [comment]
@Mothblocks I don't have the runtime on hand (this is something I fixed a long time ago on a fork), but I can explain how it would occur.

Example outfit that would trigger the runtime:
```
/datum/outfit/example_outfit
	implants = list(/obj/item/implant/uplink)
```

Outfit implant handling code: https://github.com/tgstation/tgstation/blob/master/code/datums/outfit.dm#L254C15-L257
```
		if(implants)
			for(var/implant_type in implants)
				var/obj/item/implant/I = SSwardrobe.provide_type(implant_type, H)
				I.implant(H, null, TRUE)
```

This would result in `user` (the second argument) being null, which seems to be correct behavior (as it is the game giving the user the implant, not any sort of mob)
https://github.com/tgstation/tgstation/blob/master/code/game/objects/items/implants/implant.dm#L104
In this case, the arguments of the call to `implant` would be passed to the signal handler, and well, `arguments[2]` would be null.

There are obvious alternative solutions to this problem, but they would break the assumption in code that the `user` argument would always be the one who actually implanted the target:

 1. Replacing `I.implant(H, null, TRUE)` with `I.implant(H, H, TRUE)`
 2. Something along the lines of this in `/obj/item/implant/proc/implant` (which would be more 'universal', but imo the code is butt-ugly):
```
var/list/implant_args = args.Copy()
if (LAZYLEN(implant_args) > 1)
	if (!implant_args[1])
		CRASH("Implant created without a target")
	implant_args[2] = implant_args[2] || implant_args[1]
```

I hope this explanation suffices.

**Comment by Absolucy** [comment]
Upon further testing, this doesn't runtime anymore.
I'm kind of confused as to why, but whatever, no runtime = no runtime.

**Pull request closed without merging**